### DISCUSSION TITLE
Trigger manual rebuild in JS api

### DISCUSSION THREAD

**Discussion by Kaio-Cesar** [rollup/rollup#5679] [Q&A]
Hi, new here
Having troubles creating a plugin, how does one trigger Rollup to rebuild similar to watch changes thought the api?
I'm using JS API to config it

My latest try was adding files to the watch, they appear on `this.addWatchFile`, but it just won't trigger changes

```js
const myPlugin = () => {
  return <Plugin>{
    name: 'myPlugin',
    async buildStart() {
      const files = await glob('./src/stuff/**')

      this.addWatchFile(path.resolve(__dirname, './src/stuff'))
      files.forEach(file => {
        this.addWatchFile(path.resolve(__dirname, file))
      })
    }
  }
}
```

Rollup version is 4.22

**Comment by chrisjcthomas** [comment]
Two things are likely happening here: how paths are resolved into `this.addWatchFile`, and how Rollup's watcher handles rebuild triggers.

### Path resolution
When `glob` returns matches, they're typically relative to your current working directory. Resolving them against `__dirname` will point to the wrong location if your config/plugin file lives in a subdirectory, which means Rollup watches non-existent paths and ignores file changes.

Use absolute paths from the glob call:

```js
import path from 'node:path';
import { glob } from 'glob';

const myPlugin = () => ({
  name: 'myPlugin',
  async buildStart() {
    const files = await glob('./src/stuff/**', { absolute: true });
    this.addWatchFile(path.resolve('./src/stuff'));
    for (const file of files) {
      this.addWatchFile(file);
    }
  },
});
```

### Triggering rebuilds via the JS API
`this.addWatchFile` tells Rollup's internal watcher (chokidar) to listen to filesystem change events on those paths. It doesn't trigger a build immediately by itself; it only marks them for future disk changes.

Rollup's `RollupWatcher` instance doesn't have an explicit `.rebuild()` method. If you want to trigger a rebuild programmatically from JavaScript:

1. If you're using `rollup.watch()`, touch the `mtime` of one of the watched files (or a dummy trigger file) using `fs.utimesSync`. Rollup will detect the timestamp update and trigger the build cycle.
2. If you don't actually need background filesystem watching and just want to run builds on command from your own server/script, call `rollup.rollup(inputOptions)` and `bundle.write(outputOptions)` directly whenever you want a build to run.
